File: SimilaritySearch.java
// Zhiyu Lin 300255509

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;


public class SimilaritySearch{
    public static void main(String[] args){
        String queryImageName = args[0];
        String datasetFolder = args[1];

        long startTime = System.nanoTime(); // start time


        ColorImage queryImage = new ColorImage("queryImages/" + queryImageName); // load query image
        ColorHistogram queryHistogram = new ColorHistogram(3); // color depth d = 3
        queryHistogram.setImage(queryImage);


        File datasetDir = new File(datasetFolder);
        File[] datasetFiles = datasetDir.listFiles((dir, name) -> name.endsWith(".jpg.txt")); // read histogram


        Map<String, Double> similarityData = new HashMap<>();
        
        for (int i = 0; i < datasetFiles.length; i++) {
            File file = datasetFiles[i];
            ColorHistogram datasetHistogram = new ColorHistogram(file.getAbsolutePath());
            double similarity = queryHistogram.compare(datasetHistogram);
            similarityData.put(file.getName(), similarity);
        }

        // sort similarityData
        List<Map.Entry<String, Double>> finalOuputSorted = new ArrayList<>(similarityData.entrySet());

        for (int i = 0; i < finalOuputSorted.size() - 1; i++) {
            for (int j = 0; j < finalOuputSorted.size() - i - 1; j++) {
                if (finalOuputSorted.get(j).getValue() < finalOuputSorted.get(j + 1).getValue()) {
                    Map.Entry<String, Double> temp = finalOuputSorted.get(j);
                    finalOuputSorted.set(j, finalOuputSorted.get(j + 1));
                    finalOuputSorted.set(j + 1, temp);
                }
            }
        }

        System.out.println("5 most similar images to the query image: \n");

        for (int i = 0; i < 5 && i < finalOuputSorted.size(); i++) {
            Map.Entry<String, Double> entry = finalOuputSorted.get(i);
            System.out.println("Top " + (i+1) + " = " + entry.getKey() + " | Similarity: " + entry.getValue());
        }


        // output query image histogram
        String histogramFilename = queryImageName + ".hist.txt"; 
        queryHistogram.saveHistogram(histogramFilename);
        System.out.println("\nQuery image histogram saved to: " + histogramFilename);

        
        // running time
        long endTime = System.nanoTime();
        System.out.println("\nrunning time is "+ (endTime- startTime)/1000000+" ms");
        
    }
}

File: Part_3/Final/Part_3/similaritySearch.go
// Zhiyu Lin 300255509, Yitao Cui 300345709

package main

import (
    "fmt"
    "image"
    _ "image/jpeg"
    "log"
    "os"
    "path/filepath"
    "sort"
    "sync"
    "time"
)

type Histo struct {
    Name string
    H    []int
}

func computeHistogram(imagePath string, depth int) (Histo, error) {
    file, err := os.Open(imagePath)
    if err != nil {
        return Histo{}, err
    }
    defer file.Close()

    img, _, err := image.Decode(file)
    if err != nil {
        return Histo{}, err
    }

    bounds := img.Bounds()
    histogram := make([]int, 1<<(depth*3))

    for y := bounds.Min.Y; y < bounds.Max.Y; y++ {
        for x := bounds.Min.X; x < bounds.Max.X; x++ {
            r, g, b, _ := img.At(x, y).RGBA()
            index := (((r >> 8) >> (8 - depth)) << (2 * depth)) + (((g >> 8) >> (8 - depth)) << depth) + ((b >> 8) >> (8 - depth))
            histogram[index]++
        }
    }

    return Histo{Name: filepath.Base(imagePath), H: histogram}, nil
}

func computeHistograms(imagePaths []string, depth int, hChan chan<- Histo, wg *sync.WaitGroup) {
    defer wg.Done()
    for _, imagePath := range imagePaths {
        histo, err := computeHistogram(imagePath, depth)
        if err != nil {
            log.Printf("Error computing histogram for %s: %v", imagePath, err)
            continue
        }
        hChan <- histo
    }
}

func similarity(h1, h2 []int) float64 {
    var dH1H2 float64
    var sumAllHistNumbers float64

    for i := 0; i < len(h1); i++ {
        dH1H2 += float64(min(h1[i], h2[i]))
        sumAllHistNumbers += float64(h1[i])
    }

    if sumAllHistNumbers == 0 {
        return 0
    }
    return dH1H2 / sumAllHistNumbers
}

func main() {
    if len(os.Args) != 3 {
        log.Fatalf("go run similaritySearch.go <queryImageFilename> <imageDatasetDirectory> \nEx:>> go run similaritySearch.go q00.jpg imageDataset2_15_20 ")
    }


    startTime := time.Now() // start timer

    queryImageName := os.Args[1]
    datasetDir := os.Args[2]
    depth := 3

    queryImagePath := filepath.Join("queryImages", queryImageName)

    queryHisto, err := computeHistogram(queryImagePath, depth)
    if err != nil {
        log.Fatalf("Failed to compute histogram for query image: %v", err)
    }

    datasetPath := filepath.Join(datasetDir, "*.jpg")
    filepaths, err := filepath.Glob(datasetPath)
    if err != nil {
        log.Fatalf("Failed to list dataset images: %v", err)
    }

    var K int = 1048       // set K here

    slices := make([][]string, K)
    for i, path := range filepaths {
        sliceIndex := i % K
        slices[sliceIndex] = append(slices[sliceIndex], path)
    }

    hChan := make(chan Histo, len(filepaths))
    var wg sync.WaitGroup

    for _, slice := range slices {
        wg.Add(1)
        go computeHistograms(slice, depth, hChan, &wg)
    }

    go func() {
        wg.Wait()
        close(hChan)
    }()

    similarityScores := make(map[string]float64)
    for histo := range hChan {
        similarityScores[histo.Name] = similarity(queryHisto.H, histo.H)
    }

    var ss []kv
    for k, v := range similarityScores {
        ss = append(ss, kv{k, v})
    }

    sort.Slice(ss, func(i, j int) bool {
        return ss[i].Value > ss[j].Value
    })

    endTime := time.Now() // end timer
    
    // output the Top 5 results
    fmt.Println("5 most similar images to the query image:")
    for i := 0; i < 5 && i < len(ss); i++ {
        fmt.Printf("Top %d = %s | Similarity: %.12f\n", i+1, ss[i].Key, ss[i].Value)
    }
    
    fmt.Printf("\nRunning time: %v\nK = %d", endTime.Sub(startTime), K) // output running time and K
}

type kv struct {
    Key   string
    Value float64
}
